lib/vic_helper.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+require 'uri'
+require 'base64'
+require 'oj'
+require 'openssl'
+
+module VIC
+  class Helper
+    class << self
+      def generate_url(id_attributes)
+        params = id_attributes.traits
+        params['timestamp'] = Time.now.utc.iso8601
+
+        canonical_string = Oj.dump(params)
+        params['signature'] = Helper.sign(canonical_string)
+
+        base_url = URI(Settings.vic.url)
+        base_url.query = URI.encode_www_form(params)
+        base_url.to_s
+      end
+
+      def sign(canonical_string)
+        digest = OpenSSL::Digest::SHA256.new
+        Base64.urlsafe_encode64(Helper.signing_key.sign(digest, canonical_string))
+      end
```

Comment:
NITPICK - This looks an awful lot like HMAC to me. Did you look at: https://ruby-doc.org/stdlib-2.1.0/libdoc/openssl/rdoc/OpenSSL/HMAC.html

Reply:
HMAC is by definition symmetric - i.e. shared key. This is a digital signature which gives the equivalent guarantees with public/private key which is what we want here (no shared secret with VIC team).